File: web/src/lib/__tests__/manual-endpoints.test.ts
import { describe, expect, it } from 'vitest'

import { emptyEndpoint, manualEndpoints } from '../manual-endpoints'

const good = (host: string, port: string, protocol: string) => ({
  ...emptyEndpoint(protocol),
  host,
  port,
})

describe('server settings somebody typed in', () => {
  it('sends both endpoints when both are complete', () => {
    const out = manualEndpoints(good('imap.x.jp', '993', 'imap'), good('smtp.x.jp', '465', 'smtp'))
    expect(out).toEqual({
      incoming: { host: 'imap.x.jp', port: 993, protocol: 'imap', tls: 'implicit' },
      outgoing: { host: 'smtp.x.jp', port: 465, protocol: 'smtp', tls: 'implicit' },
    })
  })

  // `Number('')` is 0, not NaN — an empty box would otherwise be sent
  // as a real port of zero, and the server would refuse it with a
  // validation error rather than the form saying what is missing.
  it('refuses an empty port rather than sending zero', () => {
    expect(
      manualEndpoints(good('imap.x.jp', '', 'imap'), good('smtp.x.jp', '465', 'smtp'))
    ).toBeNull()
  })

  it('refuses a half-filled pair', () => {
    expect(manualEndpoints(good('', '993', 'imap'), good('smtp.x.jp', '465', 'smtp'))).toBeNull()
    expect(manualEndpoints(good('imap.x.jp', '993', 'imap'), good('', '465', 'smtp'))).toBeNull()
  })

  it('refuses a port outside the range', () => {
    for (const p of ['0', '65536', '-1', '99999']) {
      expect(manualEndpoints(good('h', p, 'imap'), good('s', '465', 'smtp'))).toBeNull()
    }
  })

  it('refuses something that is not a whole number', () => {
    for (const p of ['99.5', 'abc', '9 9', '1e3']) {
      expect(manualEndpoints(good('h', p, 'imap'), good('s', '465', 'smtp'))).toBeNull()
    }
  })

  it('keeps the protocol and the encryption as chosen', () => {
    const out = manualEndpoints(
      { host: 'pop.x.jp', port: '110', protocol: 'pop3', tls: 'starttls' },
      { host: 'smtp.x.jp', port: '587', protocol: 'smtp', tls: 'starttls' }
    )
    expect(out?.incoming.protocol).toBe('pop3')
    expect(out?.incoming.tls).toBe('starttls')
  })
})

File: android/app/src/test/kotlin/jp/golia/mailrs/wire/ManualEndpointsTest.kt
package jp.golia.mailrs.wire

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class ManualEndpointsTest {
    private fun e(host: String, port: String, proto: String = "imap") =
        ManualEndpoint(host = host, port = port, proto = proto)

    @Test
    fun `both endpoints go out when both are complete`() {
        val out = wireEndpoints(e("imap.x.jp", "993"), e("smtp.x.jp", "465", "smtp"))
        assertEquals(WireEndpoint("imap.x.jp", 993, "imap", "implicit"), out?.first)
        assertEquals(WireEndpoint("smtp.x.jp", 465, "smtp", "implicit"), out?.second)
    }

    // An empty box must not become a real port. Surrounding spaces
    // are trimmed first — somebody pasting " 993 " meant 993 — but
    // anything that is not a digit after that is refused, so "+993"
    // does not arrive as a port nobody typed.
    @Test
    fun `an empty port is refused rather than sent`() {
        assertNull(wireEndpoints(e("imap.x.jp", ""), e("smtp.x.jp", "465", "smtp")))
    }

    @Test
    fun `a half-filled pair never leaves the phone`() {
        assertNull(wireEndpoints(e("", "993"), e("smtp.x.jp", "465", "smtp")))
        assertNull(wireEndpoints(e("imap.x.jp", "993"), e("", "465", "smtp")))
    }

    @Test
    fun `a port outside the range is refused`() {
        for (p in listOf("0", "65536", "99999")) {
            assertNull("port $p was accepted", wireEndpoint(e("h", p)))
        }
    }

    @Test
    fun `only digits count as a port`() {
        for (p in listOf("+993", "9 9", "99.5", "abc", "1e3")) {
            assertNull("port $p was accepted", wireEndpoint(e("h", p)))
        }
    }

    @Test
    fun `spaces around a port are somebody's paste, not a mistake`() {
        assertEquals(993, wireEndpoint(e("h", " 993 "))?.port)
    }

    @Test
    fun `the protocol and the encryption survive`() {
        val out = wireEndpoint(ManualEndpoint("pop.x.jp", "110", "pop3", "starttls"))
        assertEquals("pop3", out?.protocol)
        assertEquals("starttls", out?.tls)
    }
}
